pkg/health/membergarbagecollector/membergarbagecollector.go

```diff
@@ -0,0 +1,175 @@
+	clientSet, err := miscellaneous.GetKubernetesClientSetOrError()
+	if err != nil {
+		logger.Errorf("failed to create kubernetes clientset: %v", err)
+		return
+	}
+	mgc, err := NewMemberGarbageCollector(logger, clientSet, etcdConfig.ConnectionTimeout)
+	if err != nil {
+		logger.Errorf("failed to initialize new etcd member garbage collector: %v", err)
+		return
+	}
+	mgc.gcTimer = time.NewTimer(hconfig.MemberGCDuration.Duration)
+	defer mgc.gcTimer.Stop()
+	mgc.logger.Info("Started etcd member garbage collector")
+
+	//Create etcd client to get etcd cluster member list
+	clientFactory := etcdutil.NewFactory(*etcdConfig)
+	etcdCluster, err := clientFactory.NewCluster()
+	if err != nil {
+		mgc.logger.Errorf("Failed to create etcd cluster client: %v", err)
+	}
+	defer etcdCluster.Close()
+
+	for {
+		select {
+		case <-mgc.gcTimer.C:
+			err := mgc.RemoveSuperfluousMembers(ctx, etcdCluster)
+			if err != nil {
+				mgc.logger.Warn(err)
+			}
+			mgc.gcTimer.Reset(hconfig.MemberGCDuration.Duration)
+		case <-ctx.Done():
+			mgc.logger.Info("Stopping etcd member garbage collector")
+			return
+		}
+	}
+}
+
+// RemoveSuperfluousMembers removes a member from the etcd cluster if its corresponding pod is not present
+func (mgc *MemberGarbageCollector) RemoveSuperfluousMembers(ctx context.Context, etcdCluster etcdclient.ClusterCloser) error {
+	sts := &appsv1.StatefulSet{}
+	if err := mgc.k8sClient.Get(ctx, client.ObjectKey{
+		Namespace: mgc.podNamespace,
+		Name:      mgc.stsName,
+	}, sts); err != nil {
+		return &errors.EtcdError{
+			Message: fmt.Sprintf("Could not fetch etcd statefulset %v with error: %v", mgc.stsName, err),
+		}
+	}
+	stsSpecReplicas := *sts.Spec.Replicas
+	stsStatusReplicas := sts.Status.Replicas
+
+	listCtx, listCtxCancel := context.WithTimeout(ctx, mgc.etcdConnectionTimeout.Duration)
+	defer listCtxCancel()
+	etcdMemberListResponse, err := etcdCluster.MemberList(listCtx)
+	if err != nil {
+		return &errors.EtcdError{
+			Message: fmt.Sprintf("Could not list etcd cluster members: %v", err),
+		}
+	}
+
+	if stsSpecReplicas >= stsStatusReplicas && stsSpecReplicas >= int32(len(etcdMemberListResponse.Members)) {
+		//Return if number of replicas in sts and etcd cluster match or cluster is in scale up scenario
+		return nil
+	}
+
+	for _, member := range etcdMemberListResponse.Members {
+		memberPod := &v1.Pod{}
+		podErr := mgc.k8sClient.Get(ctx, client.ObjectKey{
+			Namespace: mgc.podNamespace,
+			Name:      member.Name,
+		}, memberPod)
+		memberLease := &coordv1.Lease{}
+		leaseErr := mgc.k8sClient.Get(ctx, client.ObjectKey{
+			Namespace: mgc.podNamespace,
+			Name:      member.Name,
+		}, memberLease)
+		if (apierrors.IsNotFound(podErr) && apierrors.IsNotFound(leaseErr)) || member.Name == "" {
+			//Remove member from etcd cluster if corresponding pod cannot be found and if corresponding member lease is not found
+			//Or remove a learner member
```

Comment:
Hm why is removing a learner required here?

Reply:
There is a possibility that the superfluous member is still in the learner state and in that case `member.Name` might also be empty if it's etcd container never started

Reply:
Got it 👍 
Is it a good idea to skip the `apierrors.IsNotFound(podErr)` in such cases? This will be a race issue as there is always a short time gap between adding the learner and starting it.

Reply:
Yes, I think we can skip the `IsNotFound` error as we will get a different error with an empty `member.Name`
I don't think we will have a race issue here as if we expect the learner to eventually start, we will not get past `stsSpecReplicas >= int32(len(etcdMemberListResponse.Members))` in line 145

Reply:
Fair enough. Thanks 🙂